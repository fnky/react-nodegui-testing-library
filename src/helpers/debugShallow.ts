import { ReactTestInstance } from 'react-test-renderer';
import shallow from '../shallow';
import format from './format';

/**
 * Log pretty-printed shallow test component instance
 */
export default function debugShallow(
  instance: ReactTestInstance | React.ReactElement<any>,
  message?: any,
): void {
  const { output } = shallow(instance);

  if (message) {
    console.log(`${message}\n\n`, format(output));
  } else {
    console.log(format(output));
  }
}
