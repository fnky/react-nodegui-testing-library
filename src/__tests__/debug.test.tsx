import * as React from 'react';
import { Button as NodeGuiButton, Text } from '@nodegui/react-nodegui';
import stripAnsi from 'strip-ansi';
import { debug, render, fireEvent, flushMicrotasksQueue } from '..';
import debugShallow from '../helpers/debugShallow';

type ConsoleLogMock = jest.MockedFunction<typeof console.log>;

function TextComponent({ text }: { text: string }): React.ReactElement {
  return <Text>{text}</Text>;
}

class Button extends React.Component<any, any> {
  state = { counter: 0 };

  onPress = async () => {
    await Promise.resolve();

    this.setState({ counter: 1 });
    this.props.onPress();
  };

  render() {
    return (
      <>
        <NodeGuiButton
          on={{ clicked: this.onPress }}
          text={`${this.props.text} ${this.state.counter}`}
        />
        <TextComponent text={`${this.props.text} ${this.state.counter}`} />
      </>
    );
  }
}

test('debug', () => {
  jest.spyOn(console, 'log').mockImplementation(x => x);
  const component = <Button onPress={jest.fn} text="Press me" />;
  debug(component);

  const output = (console.log as ConsoleLogMock).mock.calls[0][0];

  expect(stripAnsi(output)).not.toEqual(output);
  expect(stripAnsi(output)).toMatchSnapshot();

  (console.log as ConsoleLogMock).mockReset();

  debug(component, 'test message');

  expect(console.log).toHaveBeenCalledWith('test message\n\n', output);
});

test('debug.shallow', () => {
  expect(debug.shallow).toBe(debugShallow);
});

test('debug.deep', () => {
  // $FlowFixMe
  console.log = jest.fn();
  const component = <Button onPress={jest.fn} text="Press me" />;
  debug.deep(component);

  const output = (console.log as ConsoleLogMock).mock.calls[0][0];

  expect(stripAnsi(output)).not.toEqual(output);
  expect(stripAnsi(output)).toMatchSnapshot();

  (console.log as ConsoleLogMock).mockReset();

  debug.deep(component, 'test message');

  expect(console.log).toHaveBeenCalledWith('test message\n\n', output);
});

test('debug.deep async test', async () => {
  // $FlowFixMe
  console.log = jest.fn();
  const { toJSON, getByType } = render(
    <Button onPress={jest.fn} text="Press me" />,
  );

  fireEvent.click(getByType(NodeGuiButton as any));
  await flushMicrotasksQueue();

  debug.deep(toJSON());

  const output = (console.log as ConsoleLogMock).mock.calls[0][0];

  expect(stripAnsi(output)).toMatchSnapshot();
});
