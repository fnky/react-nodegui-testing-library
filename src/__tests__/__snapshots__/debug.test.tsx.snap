// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug 1`] = `
"<React.Fragment>
  <button
    on={
      Object {
        \\"clicked\\": [Function anonymous],
      }
    }
    text=\\"Press me 0\\"
  />
  <TextComponent
    text=\\"Press me 0\\"
  />
</React.Fragment>"
`;

exports[`debug.deep 1`] = `
"Array [
  <button
    on={
      Object {
        \\"clicked\\": [Function anonymous],
      }
    }
    text=\\"Press me 0\\"
  />,
  <text>
    Press me 0
  </text>,
]"
`;

exports[`debug.deep async test 1`] = `
"Array [
  <button
    on={
      Object {
        \\"clicked\\": [Function anonymous],
      }
    }
    text=\\"Press me 1\\"
  />,
  <text>
    Press me 1
  </text>,
]"
`;
